packages/blockchain-wallet-v4-frontend/src/data/goals/slice.ts

```diff
@@ -0,0 +1,130 @@
+import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit'
+import { append, assoc, filter } from 'ramda'
+
+import { ModalNameType, ModalParamPropsType } from 'data/modals/types'
+
+import { GoalsState, GoalsType } from './types'
+
+const initialState: GoalsState = {
+  goals: [],
+  initialModalDisplayed: false,
+  initialModals: {},
+  initialRedirect: ''
+}
+
+type AddInitialModalPayload = {
+  data: ModalParamPropsType
+  key: string
+  name: ModalNameType
+}
+
+type AddInitialRedirectPayload = {
+  path: string
+}
+
+type SaveGoalPayload = {
+  data
+  name: GoalsType
+}
+
+type DeleteGoalPayload = {
+  id: string
+}
+
+const generateId = () => Math.random().toString(36).substr(2, 10)
+
+const goalsSlice = createSlice({
+  extraReducers: (builder) => {
+    // Handle runGoals and defineGoals actions that does not change redux state
+    builder.addDefaultCase((state) => state)
+  },
+  initialState,
+  name: 'goals',
+  reducers: {
+    addInitialModal: {
+      prepare: (
+        key: AddInitialModalPayload['key'],
+        name: AddInitialModalPayload['name'],
+        data: AddInitialModalPayload['data']
+      ) => {
+        return { payload: { data, key, name } }
+      },
+      reducer: (state, action: PayloadAction<AddInitialModalPayload>) => {
+        return {
+          ...state,
+          initialModals: {
+            ...state.initialModals,
+            [action.payload.key]: action.payload
+          }
+        }
+      }
+    },
+    addInitialRedirect: {
+      prepare: (path: AddInitialRedirectPayload['path']) => {
+        return { payload: { path } }
+      },
+      reducer: (state, action: PayloadAction<AddInitialRedirectPayload>) => {
+        return {
+          ...state,
+          initialRedirect: action.payload.path
+        }
+      }
+    },
+    deleteGoal: {
+      prepare: (id: DeleteGoalPayload['id']) => {
+        return { payload: { id } }
+      },
+      reducer: (state, action: PayloadAction<DeleteGoalPayload>) => {
+        const { id } = action.payload
+        return assoc(
+          'goals',
+          filter((a) => a.id !== id, state.goals),
+          state
+        )
+      }
+    },
+    initialModalDisplayed: (state) => {
+      return {
+        ...state,
+        initialModalDisplayed: true
```

Comment:
```javascript
    addInitialModal: (state, action: PayloadAction<AddInitialModalPayload>) => {
      state.initialModals = {
        ...state.initialModals,
        [action.payload.key]: action.payload
      }
    },
```

Is this correct way of doing it?

Reply:
```js
    addInitialModal: (state, action: PayloadAction<AddInitialModalPayload>) => {
      state.initialModals[action.payload.key] =  action.payload
    },
 ```
 This could also be an option

Reply:
Yeah this is an interesting case. Typically you would know the key and just reference it directly but in this case you don't. I think the last example is the best one to use here as it's the simplest.